lib/uid_record.js

```diff
@@ -0,0 +1,95 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * This module keeps track of events related to a specific uid
+ */
+module.exports = function (limits, now) {
+
+  now = now || Date.now
+  function UidRecord() {
+    this.timestampsByAction = {}
+    this.rateLimitedTimestamp = undefined
+  }
+
+  UidRecord.parse = function (object) {
+    var rec = new UidRecord()
+    object = object || {}
+    rec.rateLimitedTimestamp = object.rateLimitedTimestamp // timestamp when the account was rateLimited
+    rec.timestampsByAction = object.timestampsByAction || {}  // timestamps on each kind of action
+    return rec
+  }
+
+  UidRecord.prototype.getMinLifetimeMS = function () {
+    return Math.max(
+      limits.uidRateLimitIntervalMs,
+      limits.uidRateLimitBanDurationMs
+    )
+  }
+
+  /*
+   * stores check timestamps if account is not blocked.
+   * returns remaining time until uid unblick
+   */
+  UidRecord.prototype.addCount = function (action) {
+    var timestamp = now()
+    this.trimCheckAuthenticated(action, timestamp)
+
+    if(!this.isRateLimited()){
+      this.timestampsByAction[action].push(timestamp)
+    }
+
+    return this.getPossibleRetryDelay(action)
+  }
+
+  UidRecord.prototype.trimCheckAuthenticated = function (action, now) {
```

Comment:
nit re: the naming of this function, it may be easier to follow if named more directly for what it does, such as `trimCounts` by analogy to `addCount`.